presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,272 @@
+ * limitations under the License.
+ */
+package io.prestosql.server;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.prestosql.execution.StageInfo;
+import io.prestosql.execution.StageState;
+import io.prestosql.execution.TaskInfo;
+import io.prestosql.execution.TaskManagerConfig;
+import io.prestosql.spi.QueryId;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.sql.DynamicFilters;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.optimizations.PlanNodeSearcher;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import javax.annotation.concurrent.Immutable;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.addSuccessCallback;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.prestosql.execution.StageState.FINISHED;
+import static io.prestosql.execution.StageState.RUNNING;
+import static io.prestosql.execution.StageState.SCHEDULED;
+import static io.prestosql.execution.StageState.SCHEDULING_SPLITS;
+import static io.prestosql.spi.predicate.TupleDomain.columnWiseUnion;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+
+@ThreadSafe
+public class DynamicFilterService
+{
+    private final ConcurrentHashMap<SourceDescriptor, SettableFuture<TupleDomain<String>>> dynamicFilterSummaries = new ConcurrentHashMap<>();
+    private final ConcurrentMap<QueryId, Supplier<List<StageInfo>>> queries = new ConcurrentHashMap<>();
+
+    @Inject
+    public DynamicFilterService(TaskManagerConfig taskConfig)
+    {
+        ScheduledExecutorService collectDynamicFiltersExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("DynamicFilterService"));
```

Comment:
you also need to shutdown `collectDynamicFiltersExecutor` on cleanup. You can use `@PreDestroy` annotation on cleanup method.

Reply:
right, added it now